File: 2.addtwonumbers.1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        #iterative version...
        if not l1:
            return l2
        if not l2:
            return l1
        
        carry = 0
        _l1 = l1
        _l2 = l2
        while _l1.next or _l2.next:
            if not _l1.next:
                _l1.next = ListNode(0)
            if not _l2.next:
                _l2.next = ListNode(0)
            x = (_l1.val if _l1 else 0) + (_l2.val if _l2 else 0)
            
            if x >= 10:
                _l1.next.val += 1
                x %= 10

            _l1.val = x
            
            _l1 = _l1.next
            _l2 = _l2.next
        _l1.val += _l2.val
        if _l1.val >= 10:
            _l1.next = ListNode(1)
            _l1.val %= 10
        
        
        return l1

File: 2.addtwonumbers.1.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut carry = 0;
        let mut l1m = &l1;
        let mut l2m = &l2;
        
        let mut dummy_head = Some(Box::new(ListNode::new(0)));
        let mut tail = &mut dummy_head;
        let empty = Box::new(ListNode::new(0));
        
        while (l1m.is_some() || l2m.is_some()) {
            let l1v = l1m.as_ref().unwrap_or(&empty).val;
            let l2v = l2m.as_ref().unwrap_or(&empty).val;
            let res = l1v + l2v + carry;
            let val:i32 = res % 10;
            carry = res / 10;
            
            l1m = &l1m.as_ref().unwrap_or(&empty).next;
            l2m = &l2m.as_ref().unwrap_or(&empty).next;
            
            tail.as_mut().unwrap().next = Some(Box::new(ListNode::new(val)));
            tail = &mut tail.as_mut().unwrap().next;
        }
        if carry != 0 {
            tail.as_mut().unwrap().next = Some(Box::new(ListNode::new(carry)));
            tail = &mut tail.as_mut().unwrap().next;
        }
        
        dummy_head.unwrap().next
        
    }
    
}
